docs/production.md

```diff
@@ -11,3 +12,180 @@ configurations are available:
+```
+
+Even though the array is flat, each even/odd pairing is actually a tuple:
+
+```text
+[<hmac_key_1>, <encryption_key_1>, <hmac_key_2>, <encryption_key_2>]
+```
+
+To rotate the cookie keys, generate two new keys of the correct lengths as
+specificed above and insert them **into the front** of the array. **Do not
+remove the existing values from the array**, as doing so will invalidate all
+existing sessions.
+
+```text
+[<NEW_HMAC_KEY>, <NEW_ENCRYPTION_KEY>, <hmac_key_1>, <encryption_key_1>, <hmac_key_2>, <encryption_key_2>]
+```
+
+Just as before, the new values should be base64-encoded:
+
+```sh
+export COOKIE_KEYS="c8+OD0vpvT/FrtGAtHc1nYhtkYMhjEEHCLgzuIiKJbskAbMI7bJxSnlBMKmc2AQmo8QVAViJuKoopuSuXE7tYw==,KRN9OK/lcs/uBWKQ2/1I0g9KR/iL3/MHuCn6esI02fs=,ARLaFwAqBGIkm5pLjAveJuahtCnX2NLoAUz2kCZKrScUaUkEaxHSvJLVYb5yAPCc441Cho5n5yp8jdEmy6hyig==,RLjcRZeqc07s6dh3OK4CM1POjHDZHC+usNU1w/XNTjM="
+```
+
+Upon deploying, all _new_ sessions will use these new keys. Old sessions will be
+automatically upgraded on the next visit. After a period of time that you deem
+acceptable (e.g. 30d), you can remove the older keys from the end of the array.
+
+You can use `openssl` or similar tooling to generate the secret. Note that this
+is **not** preferred since it requires a human to see the plaintext secret.
+
+```sh
+openssl rand -base64 64 | tr -d "\n" # or 32
+```
+
+If you are using a secret manager, put this value in the secret manager and
+provie its _reference_ in the environment. If you are not using a secret
+manager, provide this value directly in the environment.
+
+
+### Cross-site request forgery (CSRF) key
+
+**Recommended frequency:** 90 days, on breach
+
+To rotate the key, generate a new 32-byte key. You can use `openssl` or similar:
+
+```sh
+openssl rand -base64 32 | tr -d "\n"
+```
+
+Update the `CSRF_AUTH_KEY` environment variable and re-deploy. The system [only
+supports a single key for CSRF](https://github.com/gorilla/csrf/issues/65). When
+you deploy the new key, any existing open HTML forms will fail to submit as an
+invalid request.
+
+
+### Database encryption keys
+
+**Recommend frequency:** 30 days, on breach
+
+These keys control application-layer encryption of secrets before they are
+stored in the database. For example, this key encrypts Twilio credentials so
+they are not in plaintext in the database. Note that we only use the encryption
+key where encryption is appropriate. For API keys and tokens, we HMAC the values
+as their plaintext values are not required.
+
+To rotate the encryption keys, rotate them in the underlying key manager. Note
+that old entries will still be encrypted with the old key. You do not need to
+upgrade them so long as the older key version is still available in your key
+manager.
+
+While unlikely, this may require you to update the `DB_ENCRYPTION_KEY`
+environment variable.
+
+
+### API Key Signature HMAC keys
+
+**Recommended frequency:** 90 days
```

Comment:
not on breach?

Reply:
It doesn't really help on breach. In the case of API keys + breach, there will need to be a business decision about how to move forward. Either break existing clients and do a full key rotation, or tune down the rate limiter to prevent abuse until you can push out new API keys to clients.